src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -1599,9 +1606,19 @@ private boolean bootstrap(final Collection<Token> tokens)
         ListenableFuture<StreamState> bootstrapStream = bootstrapper.bootstrap(streamStateStore, !replacing && useStrictConsistency); // handles token update
         try
         {
-            bootstrapStream.get();
+            while (true) {
+                if (bootstrapStream.isDone()) {
+                    bootstrapStream.get();
+                    logger.info("Bootstrap streaming completed for tokens {}", tokens);
+                    break;
+                }
+                if (hasNonTransientError(NonTransientError.BOOTSTRAP_ERROR)) {
+                    logger.info("Stopped waiting for bootstrap streaming to complete because detected a bootstrap error.", SafeArg.of("nonTransientErrors", getNonTransientErrors()));
+                    break;
+                }
+                Uninterruptibles.sleepUninterruptibly(1, TimeUnit.SECONDS);
+            }
```

Comment:
You might be able to do less aggressive polling while keeping the reactivity by doing something like
````
while (true) {
    try {
        bootstrapStream.get(timeBetweenCheck, TimeUnit.SECONDS);
        logger.info("Bootstrap streaming completed for tokens {}", tokens);
        break;
    } catch (TimeoutException e) {
        if (hasNonTransientError(NonTransientError.BOOTSTRAP_ERROR)) {
            logger.info("Aborting waiting for bootstrap streaming to complete because detected a bootstrap error.",
                    SafeArg.of("nonTransientErrors", getNonTransientErrors()));
            break;
        }
    } catch (Exception e) {
        logger.error("Error while waiting for bootstrap stream to complete", e);
        break;
    }
}
```
If we stop waiting for the streams, is there a way to also notify peers/shut down stream early? Or will the following `disableNode` handle this

Reply:
this just blocks on `bootstrapStream.get`, right? like the `while(true)` doesn't do anything here

Reply:
hm, why is this less aggressive polling? if possible, i'd like to avoid using exceptions for control flow and i think i don't recognize the benefit here

Reply:
> notify peers/shut down stream early

i'll test this tomorrow to see if `disableNode` kills the streams as we would hope. with operator, the node will be deleted after entering NTE anyway though

Reply:
`disableNode` will not kill the streams. i just realized that in the original instance that this happened, the streams continued even after the NTE blip